python/ray/experimental/data/impl/reader.py

```diff
@@ -21,9 +23,53 @@ def read_file(path: str,
             filesystem = filesystem.unwrap()
         contents = filesystem.open_input_stream(path).readall()
     else:
-        contents = open(path, "rb").read()
+        if path.startswith("s3://"):
+            contents = s3_downloader(path)
+        else:
+            contents = open(path, "rb").read()
 
     if include_paths:
         return path, contents
     else:
         return contents
+
+
+download_initialized = False
+s3_client = None
+http_session = None
+
+
+def s3_downloader(path: str) -> bytes:
+    # S3 specific dependencies should only be imported if we're downloading a
+    # file from s3.
+    import boto3
+    import requests
+
+    global download_initialized, s3_client, http_session
+    if download_initialized is False:
+        s3_client = boto3.client("s3")
+        http_session = requests.Session()
+        download_initialized = True
+
+    without_prefix = path[len("s3://"):]
+    bucket, key = without_prefix.split("/", 1)
+
+    # TODO (Alex): We should consider generating a presigned url once we port
+    # this to the file system abstraction.
+    # url = s3_client.generate_presigned_url(
+    #     "get_object",
+    #     Params={"Bucket": bucket, "Key": key},
+    #     ExpiresIn=3600
+    # )
+    url = f"https://{bucket}.s3.amazonaws.com/{key}"
+
+    # Retry download if it fails.
+    for _ in range(3):
+        result = http_session.get(url)
+        if result.status_code == 200:
+            break
+        logger.warning(
+            f"Failed to download {url} with error: {result.content}. Retrying."
+        )
+
```

Comment:
Raise properly on error

Reply:
I think we want a warning. S3 just errors sometimes. The nightly test won't consistently pass without a retry.

Reply:
I mean raise if it fails 3 times